src/mds/MDCache.cc

```diff
@@ -9812,6 +9812,19 @@ void MDCache::dispatch_request(const MDRequestRef& mdr)
 {
   if (mdr->dead) {
     dout(20) << __func__ << ": dead " << *mdr << dendl;
+    //if the mdr is a "batch_op" and it has followers, pick a follower as
+    //the new "head of the batch ops" and go on processing the new one.
+    if (mdr->client_request && mdr->is_batch_head()) {
+      int mask = mdr->client_request->head.args.getattr.mask;
+      auto it = mdr->batch_op_map->find(mask);
+      auto new_batch_head = it->second->find_new_head();
+      if (!new_batch_head) {
+	mdr->batch_op_map->erase(it);
+	return;
+      }
+      mds->finisher->queue(new C_MDS_RetryRequest(this, new_batch_head));
```

Comment:
i have some questions - from what i understand is that if we have a dead request coming in and if its a batch_ops head then we first try to locate it in the `batch_op_map` and try to get a new head then there is a `if {}` that checks if we have a valid ref to the new head else we delete the dead head's entry from the map, but if this happens how do we get a new batch head? using `mds->finisher->queue(new C_MDS_RetryRequest(this, new_batch_head));` ? but the `new_batch_head` is `nullptr` in this case so does this line have any impact? if there is  no `new_batch_head` then shouldn't we pick the `it` +1 as the new head? and only if `new_batch_head != nullptr`  we should `ds->finisher->queue(new C_MDS_RetryRequest(this, new_batch_head));` ?

Reply:
```
 4091     } else if (is_lookup) {
 4092       CDentry* dn = mdr->dn[0].back();    
 4093       mdr->pin(dn);
 4094       auto em = dn->batch_ops.emplace(std::piecewise_construct, std::forward_as_tuple(mask), std::forward_as_tuple());
 4095       if (em.second) {
 4096         em.first->second = std::make_unique<Batch_Getattr_Lookup>(this, mdr);  
 4097       } else {
 4098         dout(20) << __func__ << ": LOOKUP op, wait for previous same getattr ops to respond. " << *mdr << dendl;
 4099         em.first->second->add_request(mdr);
 4100         mdr->mark_event("joining batch lookup");
 4101         return;
 4102       }
           
```

From the above code we can see that when the first `lookup` request comes it will mark it as a batch head and add it to the `dn's batch_op<int , std::unique_ptr<BatchOp>>` map. Please note the batch head one will be a `Batch_Getattr_Lookup` and for the followers will be added to `Batch_Getattr_Lookup`.

For example:

`dn --> batch_op['mask_X']--> Batch_Getattr_Lookup : {'mdr', 'batch_reqs<>'}`

And the 'mdr' will be batch head and the `batch_reqs<>` will be a map of the followers having the same `mask_X`.

When the batch head 'mdr' is dead or killed the mds will choose a new `mdr_X` from the `batch_reqs<>` map and replace the batch head 'mdr' in `Batch_Getattr_Lookup`. And finally queue the `mdr_X` as the batch head.

When replacing the batch head `mdr` the reference of the old `mdr` will be release, `**this PR is trying to fix this**`.

But if there has no any follower, then the `new_batch_head` will be `nullptr` and then there is no need to queue it. In this case we will do nothing and will just erase the `Batch_Getattr_Lookup` struct and `return` it.

Is this explain it ?

Reply:
yeah, btw your recent code is what i was talking about, previously the request was being queued regardless of it being a valid ref, but now it does :)